### PR TITLE
resources

### PR DESCRIPTION
**Pull request by dima-avdeev-jb**
Usage:

```Kotlin
Image( resource("dir/img.png").rememberImageBitmap().orEmpty() )
```

### CODE REVIEW

File: components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+@Composable
+fun CountRecompositions(imageBitmap: ImageBitmap?, onRecomposition: () -> Unit) {
+    onRecomposition()
+    println("imageBitmap: $imageBitmap")
+    if (imageBitmap != null) {
+        Image(bitmap = imageBitmap, contentDescription = null)
+    }
+}
```
Comment (dima-avdeev-jb) [inline components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt:59]: Here test not working

File: components/gradle.properties [hunk 2]
```diff
@@ -5,4 +5,18 @@ kotlin.code.style=official
 # __KOTLIN_COMPOSE_VERSION__
 kotlin.version=1.7.20
 # __LATEST_COMPOSE_RELEASE_VERSION__
-compose.version=1.2.1
+compose.version=1.3.0-beta02
```
Comment (dima-avdeev-jb) [inline components/gradle.properties:8]: 1.3.0-beta02 need for Android tests

File: components/resources/demo/shared/src/commonMain/kotlin/org/jetbrains/compose/resources/demo/shared/UseResources.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+@Composable
+internal fun UseResources() {
+    Column {
+        Text("Hello, resources")
+        Image(
+            bitmap = resource("dir/img.png").rememberImageBitmap().orEmpty(),
+            contentDescription = null,
+        )
+        Image(
+            bitmap = resource("img.webp").rememberImageBitmap().orEmpty(),
```
Comment (dima-avdeev-jb) [inline components/resources/demo/shared/src/commonMain/kotlin/org/jetbrains/compose/resources/demo/shared/UseResources.kt:19]: Also webp supported

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/ComposeResource.common.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    return state.value
+}
+
+/**
+ * return current ImageBitmap or return empty while loading
+ */
+@ExperimentalResourceApi
+@Composable
+inline fun ImageBitmap?.orEmpty(): ImageBitmap = this ?: remember { ImageBitmap(1, 1) }
```
Comment (eymar) [inline components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/ComposeResource.common.kt:35]: Maybe we can have a singleton empty ImageBitmap instance?

```kotlin
private val emptyImageBitmap: ImageBitmap by lazy { ImageBitmap(1, 1) }

// an then:
@ExperimentalResourceApi
// and now no need to make it @Composable
inline fun ImageBitmap?.orEmpty(): ImageBitmap = this ?: emptyImageBitmap
```

 I'm not sure if the same ImageBitmap instance can be reused in different places. But it seems there should be no such a limitation.

File: components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        var recompositionCount = 0
+        rule.setContent {
+            val state: Boolean by mutableStateFlow.collectAsState(true)
+            val resource = resource(if (state) "1.png" else "2.png")
+            CountRecompositions(resource.rememberImageBitmap()) {
+                recompositionCount++
+            }
+        }
+        rule.awaitIdle()
+        delay(10)
```
Comment (eymar) [inline components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt:48]: doesn't it work without a delay(10)?

File: components/resources/library/src/macosMain/kotlin/org/jetbrains/compose/resources/Resource.macos.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+@ExperimentalResourceApi
+actual fun resource(path: String): Resource = MacOSResourceImpl(path)
+
+@ExperimentalResourceApi
+private class MacOSResourceImpl(val path: String) : Resource {
+    override suspend fun readBytes(): LoadState<ByteArray> {
+        val currentDirectoryPath = NSFileManager.defaultManager().currentDirectoryPath
+        val contentsAtPath: NSData? = NSFileManager.defaultManager().run {
+            contentsAtPath("$currentDirectoryPath/src/macosMain/resources/$path")
```
Comment (eymar) [inline components/resources/library/src/macosMain/kotlin/org/jetbrains/compose/resources/Resource.macos.kt:22]: The hierarchy of source sets can be customized (for example like we have in androidx fork: commonMain -> skikoMain -> nativeMain -> macosMain).

I think for now we can limit the resource directory to be in macosMain or commonMain. Maybe leaving a comment or TODO about custom sourceSets might be helpful in the future.

File: components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt (hunk 5 above)
Comment (dima-avdeev-jb) [reply components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt:48]: Yeah, it works!

File: components/resources/library/src/macosMain/kotlin/org/jetbrains/compose/resources/Resource.macos.kt (hunk 6 above)
Comment (dima-avdeev-jb) [reply components/resources/library/src/macosMain/kotlin/org/jetbrains/compose/resources/Resource.macos.kt:23]: Done

File: components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt (hunk 5 above)
Comment (dima-avdeev-jb) [reply components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt]: Done

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/ComposeResource.common.kt (hunk 4 above)
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/ComposeResource.common.kt]: Done

File: components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+class ComposeResourceTest {
+
+    @get:Rule
+    val rule = createComposeRule()
+
+    @Test
+    fun testMissingResource() = runTest (UnconfinedTestDispatcher()) {
+        var recompositionCount = 0
+        rule.setContent {
+            CountRecompositions(resource("missing.png").rememberImageBitmap()) {
```
Comment (igordmn) [inline components/resources/library/src/desktopTest/kotlin/org/jetbrains/compose/resources/ComposeResourceTest.kt:28]: To continue our conversation about ability to load resources the blocking way. I imagine something like this:

```
BlockingLoader {

}

AsyncLoader(
  loading = { ProgressBar() }, // default { }
  error = { }, // default { }
) {
    ...
    Image(resource("img.png").rememberImageBitmap())
    ...
}

interface LoadQueue {
    fun <T> add(task: Deferred<T>)
}

val LocalLoadQueue: CompositionLocal<Loader>

// adds a load to queue. when loaded inform the parent container that we can show the content.
// Not sure, but probably we can't return ImageBitmap here. Only something that access image on a later stage.
@Composable
fun Resource.rememberImageBitmap(): Painter
```

But I don't know what to do with fonts. There is nothing like Painter there.

So, this is a big area for thoughts. Let's postpone it to the later and merge only the async version (when we resolve the minor discussions).

File: components/resources/library/build.gradle.kts [hunk 8]
```diff
@@ -1,24 +1,114 @@
...
+                    device = "Pixel 5"
+                    apiLevel = 31
+                    systemImageSource = "aosp"
+                }
+            }
+        }
+    }
+}
+
+dependencies {
```
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/build.gradle.kts:105]: androidTest is not available in Kotlin multiplatform sourcceSet configuration

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+package org.jetbrains.compose.resources
+
+/**
+ * Should implement equals() and hashCode()
+ */
+@ExperimentalResourceApi
+interface Resource {
+    suspend fun readBytes(): LoadState<ByteArray> //todo in future use streaming
```
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt:13]: Changed to Result<ByteArray>
I will think about Exception's

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+/**
+ * Get resource from sourceSet/resources.
+ * For examples commonMain/resources
+ */
+@ExperimentalResourceApi
+expect fun resource(path: String): Resource
+
+internal class MissingResource(path: String):Exception("missing resource with path: $path")
```
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt]: Done

[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt]: IOException is available only for desktopMain and androidMain. I created expect/actual class MissingResourceException

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/ComposeResource.common.kt [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.ui.graphics.ImageBitmap
+
+private val emptyImageBitmap: ImageBitmap by lazy { ImageBitmap(1, 1) }
+
+/**
+ * Get and remember resource. While loading and if resource not exists result will be null.
```
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/ComposeResource.common.kt:17]: Yeah, good suggestion!

[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/ComposeResource.common.kt:14]: Done

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt]: Also, changed to Exception's
But we can't use IOException it's Java-specific.

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt [hunk 12]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright 2020-2022 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE.txt file.
+ */
+
+package org.jetbrains.compose.resources
+
+/**
+ * Should implement equals() and hashCode()
```
Comment (pjBooms) [inline components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt:9]: As it is public API, I believe we should write more about this abstract interface.

File: components/README.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
+# Libraries for Compose Multiplatform
```
Comment (pjBooms) [inline components/README.md:1]: Can we follow general KKM project structure?

File: components/resources/library/src/androidMain/kotlin/org/jetbrains/compose/resources/Resource.android.kt [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+package org.jetbrains.compose.resources
+
+@ExperimentalResourceApi
+actual fun resource(path: String): Resource = AndroidResourceImpl(path)
+
+@ExperimentalResourceApi
+private class AndroidResourceImpl(val path: String) : Resource {
+    override suspend fun readBytes(): LoadState<ByteArray> {
+        val contextClassLoader = Thread.currentThread().contextClassLoader!!
```
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/androidMain/kotlin/org/jetbrains/compose/resources/Resource.android.kt:16]: yes, it works in demo project

File: components/resources/library/src/androidMain/kotlin/org/jetbrains/compose/resources/Resource.android.kt [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE.txt file.
+ */
+
+package org.jetbrains.compose.resources
+
+@ExperimentalResourceApi
+actual fun resource(path: String): Resource = AndroidResourceImpl(path)
+
+@ExperimentalResourceApi
+private class AndroidResourceImpl(val path: String) : Resource {
```
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/androidMain/kotlin/org/jetbrains/compose/resources/Resource.android.kt:14]: I will move equals & hashcode to abstract class

File: components/resources/library/build.gradle.kts [hunk 16]
```diff
@@ -1,24 +1,114 @@
...
     id("maven-publish")
+    id("com.android.library")
 }
 
+val composeVersion = extra["compose.version"] as String
+
 kotlin {
     jvm("desktop")
+    android {
+        publishLibraryVariants("release", "debug")
```
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/build.gradle.kts:15]: Yes, we need it for library publication

File: components/README.md (hunk 13 above)
Comment (dima-avdeev-jb) [reply components/README.md:1]: Yes

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt (hunk 12 above)
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt:9]: Sure, we can do it later

File: components/resources/library/src/androidMain/kotlin/org/jetbrains/compose/resources/Resource.android.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/androidMain/kotlin/org/jetbrains/compose/resources/Resource.android.kt]: Changed to ***ResourceImpl.javaClass.classLoader

File: components/README.md (hunk 13 above)
Comment (dima-avdeev-jb) [reply components/README.md:1]: I will change structure

Comment (dima-avdeev-jb) [reply components/README.md:1]: Almost done.
Leave JS for not as it was.

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (igordmn) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt]: > But we can't use IOException it's Java-specific.

Better then to catch only `Exception`, not `Throwable`. We don't want to catch OutOfMemory here, for example.

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (igordmn) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt]: Ok, let's keep it that way (as implementation details of the resource system for each platform)

nit: it is better to make `missing` with capital `M` :)

File: components/resources/library/build.gradle.kts (hunk 16 above)
[the remark this answers is not in the document]
Comment (igordmn) [reply components/resources/library/build.gradle.kts:15]: We need only `release`. No need for additional `debug` artifact. It should work (it works in Compose itself)

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt]: Done

File: components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/src/commonMain/kotlin/org/jetbrains/compose/resources/Resource.common.kt]: Done

File: components/resources/library/build.gradle.kts (hunk 16 above)
[the remark this answers is not in the document]
Comment (dima-avdeev-jb) [reply components/resources/library/build.gradle.kts:15]: Done

**Review by igordmn — APPROVED** [review]

**Pull request merged**